scipy/spatial/src/distance_impl.h

```diff
@@ -220,71 +220,71 @@ russellrao_distance_char(const char *u, const char *v, npy_intp n)
 static NPY_INLINE double
 kulsinski_distance_char(const char *u, const char *v, npy_intp n)
 {
-    npy_intp _i;
-    npy_intp ntt = 0, nft = 0, ntf = 0, nff = 0;
+    npy_intp i;
+    npy_intp ntt = 0, ndiff = 0;
 
-    for (_i = 0; _i < n; _i++) {
-        ntt += (u[_i] && v[_i]);
-        ntf += (u[_i] && !v[_i]);
-        nft += (!u[_i] && v[_i]);
-        nff += (!u[_i] && !v[_i]);
+    for (i = 0; i < n; i++) {
+        npy_bool x = u[i], y = v[i];
+        ntt += x & y;
+        ndiff += (x != y);
     }
-    return (double)(ntf + nft - ntt + n) / (double)(ntf + nft + n);
+    return ((double)ndiff - ntt + n) / ((double)ndiff + n);
 }
 
 static NPY_INLINE double
 sokalsneath_distance_char(const char *u, const char *v, npy_intp n)
 {
-    npy_intp _i;
-    npy_intp ntt = 0, nft = 0, ntf = 0;
+    npy_intp i;
+    npy_intp ntt = 0, ndiff = 0;
 
-    for (_i = 0; _i < n; _i++) {
-        ntt += (u[_i] && v[_i]);
-        ntf += (u[_i] && !v[_i]);
-        nft += (!u[_i] && v[_i]);
+    for (i = 0; i < n; i++) {
+        npy_bool x = u[i], y = v[i];
+        ntt += x & y;
+        ndiff += (x != y);
     }
-    return (2.0 * (ntf + nft)) / (2.0 * (ntf + nft) + ntt);
+    return (2. * ndiff) / (2. * ndiff + ntt);
 }
 
 static NPY_INLINE double
 sokalmichener_distance_char(const char *u, const char *v, npy_intp n)
 {
     npy_intp i;
-    npy_intp ntt = 0, nft = 0, ntf = 0, nff = 0;
+    npy_intp ntt = 0, ndiff = 0;
 
     for (i = 0; i < n; i++) {
-        ntt += (u[i] && v[i]);
-        nff += (!u[i] && !v[i]);
-        ntf += (u[i] && !v[i]);
-        nft += (!u[i] && v[i]);
+        npy_bool x = u[i], y = v[i];
+        ntt += x & y;
+        ndiff += (x != y);
     }
-    return (2.0 * (ntf + nft)) / (2.0 * (ntf + nft) + ntt + nff);
+    return (2. * ndiff) / ((double)ndiff + n);
 }
 
 static NPY_INLINE double
 jaccard_distance_double(const double *u, const double *v, npy_intp n)
 {
-    double denom = 0.0, num = 0.0;
+    npy_intp denom = 0, num = 0;
     npy_intp i;
 
     for (i = 0; i < n; i++) {
-        num += (u[i] != v[i]) & ((u[i] != 0.0) | (v[i] != 0.0));
-        denom += (u[i] != 0.0) | (v[i] != 0.0);
+        double x = u[i], y = v[i];
+        num += (x != y) & ((x != 0.0) | (y != 0.0));
+        denom += (x != 0.0) | (y != 0.0);
     }
-    return num / denom;
+    return (double)num / denom;
 }
 
 static NPY_INLINE double
 jaccard_distance_char(const char *u, const char *v, npy_intp n)
 {
-    double num = 0.0, denom = 0.0;
+    npy_intp num = 0.0, denom = 0.0;
     npy_intp i;
 
     for (i = 0; i < n; i++) {
-        num += (u[i] != v[i]) & ((u[i] != 0) | (v[i] != 0));
-        denom += (u[i] != 0) | (v[i] != 0);
+        npy_bool x = u[i], y = v[i];
+        num += (x != y) & ((x != 0) | (y != 0));
+        denom += (x != 0) | (y != 0);
```

Comment:
This could simply be `x | y` as they are booleans, wonder if it makes a performance difference, probably not.